README
NAME
    Plack::Middleware::Proxy::Requests - Forward proxy server

SYNOPSIS
      # In app.psgi
      use Plack::Builder;
      use Plack::App::Proxy;

      builder {
          enable "Proxy::Connect";
          enable "Proxy::AddVia";
          enable "Proxy::Requests";
          Plack::App::Proxy->new->to_app;
      };

      # From shell
      plackup -s Starlet -E Proxy -e 'enable q{AccessLog}' app.psgi

      # or
      plackup -s Starlet -E Proxy -MPlack::App::Proxy \
              -e 'enable q{AccessLog}; enable q{Proxy::Connect}; \
                  enable q{Proxy::AddVia}; enable q{Proxy::Requests}; \
                  Plack::App::Proxy->new->to_app'

DESCRIPTION
    This module handles HTTP requests as a forward proxy server.

    Its job is to set a `plack.proxy.url' environment variable based on
    `REQUEST_URI' variable.

    The HTTP responses from the Internet might be invalid. In that case it
    is better to run the server without Plack::Middleware::Lint module. This
    module is started by default and disabled if `-E' option is used when
    starting plackup script. Note that this disable also
    Plack::Middleware::AccessLog so it have to be enabled explicitly if
    needed.

    In some cases the default server Plack::Server::PSGI alias `Standalone'
    can hang up. It is better to run proxy server with Starlet, Starman or
    Twiggy.

SEE ALSO
    Plack, Plack::App::Proxy, Plack::Middleware::Proxy::Connect,
    Plack::Middleware::Proxy::AddVia, Starlet, Starman, Twiggy.

BUGS
    If you find the bug or want to implement new features, please report it
    at https://github.com/dex4er/perl-Plack-Middleware-Proxy-Requests/issues

    The code repository is available at
    http://github.com/dex4er/perl-Plack-Middleware-Proxy-Requests

AUTHOR
    Piotr Roszatycki <dexter@cpan.org>

LICENSE
    Copyright (c) 2012-2013 Piotr Roszatycki <dexter@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as perl itself.

    See http://dev.perl.org/licenses/artistic.html



Changes
0.0103  2023-06-17      Piotr Roszatycki <dexter@cpan.org>

* Provides cpanfile.

0.0102  2013-05-24      Piotr Roszatycki <dexter@cpan.org>

* Prefer Twiggy as a server (after long testing it seems to be the best
  choice)

0.0101  2013-05-23      Piotr Roszatycki <dexter@cpan.org>

* Prefer Starlet as a server.
* Added an example for caching proxy server.

0.01    2012-08-14      Piotr Roszatycki <dexter@cpan.org>

* First CPAN release.


Build.PL
#!/usr/bin/perl

use 5.006;

use strict;
use warnings;

use Module::Build;
use Module::CPANfile;

my $module = 'Plack::Middleware::Proxy::Requests';

(my $dist = $module) =~ s/::/-/g;

my $prereqs = Module::CPANfile->load->prereqs;

my $build = Module::Build->new(
    module_name             => $module,
    license                 => 'perl',
    build_requires          => $prereqs->requirements_for('build', 'requires')->as_string_hash,
    requires                => $prereqs->requirements_for('runtime', 'requires')->as_string_hash,
    recommends              => $prereqs->requirements_for('runtime', 'recommends')->as_string_hash,
    test_requires           => $prereqs->requirements_for('test', 'requires')->as_string_hash,
    auto_configure_requires => 0,
    create_makefile_pl      => 'traditional',
    create_readme           => 1,
    create_license          => 1,
    recursive_test_files    => 0,
    add_to_cleanup          => [
        '*.bak', '*.gz', '*.tgz', '*.ppd', '*.ppmx', '*.deb', '*.bat',
        'Makefile', 'pm_to_blib', 'cover_db', 'Debian_CPANTS.txt',
        "$dist-*",
    ],
    meta_merge => {
        resources => {
            bugtracker => "http://github.com/dex4er/perl-$dist/issues",
            repository => "http://github.com/dex4er/perl-$dist",
        },
    },
);

$build->create_build_script;


lib/Plack/Middleware/Proxy/Requests.pm
package Plack::Middleware::Proxy::Requests;

=head1 NAME

Plack::Middleware::Proxy::Requests - Forward proxy server

=head1 SYNOPSIS

=for markdown ```perl

    # In app.psgi
    use Plack::Builder;
    use Plack::App::Proxy;

    builder {
        enable "Proxy::Connect";
        enable "Proxy::AddVia";
        enable "Proxy::Requests";
        Plack::App::Proxy->new->to_app;
    };

=for markdown ```

=for markdown ```sh

    # From shell
    plackup -s Twiggy -E Proxy -e 'enable q{AccessLog}' app.psgi

    # or
    twiggy -MPlack::App::Proxy \
            -e 'enable q{AccessLog}; enable q{Proxy::Connect}; \
                enable q{Proxy::AddVia}; enable q{Proxy::Requests}; \
                Plack::App::Proxy->new->to_app'

=for markdown ```

=head1 DESCRIPTION

This module handles HTTP requests as a forward proxy server.

Its job is to set a C<plack.proxy.url> environment variable based on
C<REQUEST_URI> variable.

The HTTP responses from the Internet might be invalid. In that case it
is required to run the server without L<Plack::Middleware::Lint> module.
This module is started by default and disabled if C<-E> or
C<--no-default-middleware> option is used when starting L<plackup>
script. Note that this disables also L<Plack::Middleware::AccessLog> so
it has to be enabled explicitly if needed.

The default server L<Plack::Server::PSGI> alias C<Standalone> can hang
up on the stalled connection. It is better to run a proxy server with
L<Starlet>, L<Starman> or L<Twiggy>.

=for readme stop

=cut

use 5.006;

use strict;
use warnings;

our $VERSION = '0.0103';

use parent qw(Plack::Middleware);

sub call {
    my ($self, $env) = @_;

    $env->{'plack.proxy.url'} = $env->{REQUEST_URI};

    return $self->app->($env);
}

1;

__END__

=for readme continue

=head1 SEE ALSO

L<Plack>, L<Plack::App::Proxy>, L<Plack::Middleware::Proxy::Connect>,
L<Plack::Middleware::Proxy::AddVia>, L<Starlet>, L<Starman>, L<Twiggy>.

=head1 BUGS

If you find the bug or want to implement new features, please report it at
L<https://github.com/dex4er/perl-Plack-Middleware-Proxy-Requests/issues>

The code repository is available at
L<http://github.com/dex4er/perl-Plack-Middleware-Proxy-Requests>

=head1 AUTHOR

Piotr Roszatycki <dexter@cpan.org>

=head1 LICENSE

Copyright (c) 2012-2013, 2023 Piotr Roszatycki <dexter@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as perl itself.

See L<http://dev.perl.org/licenses/artistic.html>
